### PR TITLE
fix: enable tls for the tcp address

### PR DESCRIPTION
**Pull request by cr7258**
we need to [enable tls](https://apisix.apache.org/docs/apisix/stream-proxy/#accept-tls-over-tcp-connection) in APISIX config file to support SNI based TLS route.  fix issue: #348
```yaml
apisix:
  stream_proxy: # TCP/UDP proxy
    tcp: # TCP proxy address list
      - addr: 9100
        tls: true
```

Add host field in [apisixroutes CRD](https://github.com/apache/apisix-ingress-controller/pull/1051/files#diff-df521c61cf5f2cbf22dc2b191f7699e255b89e9408b3eb02385efe733fd39897) to support SNI based TLS route. 
```yaml
apiVersion: apisix.apache.org/v2
kind: ApisixRoute
metadata:
  name: redis-stream
spec:
  stream:
  - name: redis-stream
    protocol: TCP
    match:
      ingressPort: 10000
      host: test.redis.com # sni
    backend:
      serviceName: redis1
      servicePort: 6379
  - name: redis-stream-2
    protocol: TCP
    match:
      ingressPort: 10000
      host: test.redis-2.com
    backend:
      serviceName: redis2
      servicePort: 6379
```

### CODE REVIEW

**Review by tao12345666333 — COMMENTED** [review]
thanks for your contribution

However, please note that this feature has been introduced since APISIX Ingress v1.6.
But the v1.6 version has not been released, it is expected next week

**Comment by cr7258** [comment]
I know this feature was implemented in [this PR](https://github.com/apache/apisix-ingress-controller/pull/1051) for apisix ingress, but apisix doesn't support to enable tls in helm chart now,  we only can set tcp port in the `stream_proxy.tcp` field in the values.yaml file.
```yaml
apisix:
  stream_proxy:
    tcp:
      -  9100
      -  9200
```
if we want to use this feature in apisix ingress, we must enable tls in apisix first. [So we should have the helm templates support to enable tls](https://github.com/apache/apisix-helm-chart/pull/401/files).  like this:
```yaml
apisix:
  stream_proxy:
    tcp:
      - addr: 9100
        tls: true  # enable tls for tcp 9100
      - 9200
```

**Comment by cr7258** [comment]
maybe I needn't change apisixroute CRD in [this commit](https://github.com/apache/apisix-helm-chart/pull/401/commits/71ecc8f98c2b90d1137c23112ddfe80726e87773)?

File: charts/apisix-ingress-controller/crds/customresourcedefinitions.yaml [hunk 1]
```diff
@@ -1782,6 +1782,8 @@ spec:
                     match:
                       type: object
                       properties:
+                        host:
```
Comment (tokers) [inline charts/apisix-ingress-controller/crds/customresourcedefinitions.yaml:1785]: Also update the definition in apisix-ingress-controller. Cc @tao12345666333

**Review by tokers — COMMENTED** [review]
Basically LGTM, could you also update the docs to introduce the new field?

File: charts/apisix-ingress-controller/crds/customresourcedefinitions.yaml (hunk 1 above)
Comment (tao12345666333) [reply charts/apisix-ingress-controller/crds/customresourcedefinitions.yaml:1785]: It's OK.

**Comment by cr7258** [comment]
> Basically LGTM, could you also update the docs to introduce the new field?

@tokers, I add a document to introduce how to use this field: [docs: add match stream route with SNI tutorial](https://github.com/apache/apisix-ingress-controller/pull/1543).

**Comment by tao12345666333** [comment]
Great! I will add this one to my list.

**Comment by tao12345666333** [comment]
@cr7258 could you please resolve conflicts? thanks!

**Comment by cr7258** [comment]
> @cr7258 could you please resolve conflicts? thanks!

Done, Please review it, Thanks.

File: charts/apisix/templates/_pod.tpl [hunk 2]
```diff
@@ -62,7 +62,11 @@ spec:
         {{- if (gt (len .tcp) 0) }}
         {{- range $index, $port := .tcp }}
         - name: proxy-tcp-{{ $index | toString }}
+        {{- if kindIs "map" $port }}
+          containerPort: {{ $port.addr }}
```
Comment (tokers) [inline charts/apisix/templates/_pod.tpl:66]: `addr` might contains IP, so here we cannot treat it as a port, we may need to use `splitList` to fetch the port from the `addr`.

Comment (tao12345666333) [reply charts/apisix/templates/_pod.tpl:66]: @cr7258 PTAL

**Comment by cr7258** [comment]
Use [`splitList`](https://pkg.go.dev/path/filepath#SplitList) to split a string into a list of strings, and get the last value of the list.

# Verify
## values.yaml
```yaml
gateway:
  stream:
    enabled: true
    only: false
    tcp:
      - addr: 9100
        tls: true
      - addr: 127.0.0.1:9200
      - addr: 127.0.0.1:9300
        tls: true
```

Install apisix helm chart.
```bash
helm install apisix -n apisix apisix
```


## Generated Config

### Deployment
```bash
> kubectl get deployments -n apisix apisix -o yaml

apiVersion: apps/v1
kind: Deployment
metadata:
  name: apisix
  namespace: apisix
......
        - containerPort: 9100
          name: proxy-tcp-0
          protocol: TCP
        - containerPort: 9200
          name: proxy-tcp-1
          protocol: TCP
        - containerPort: 9300
          name: proxy-tcp-2
          protocol: TCP
......
```

### Configmap
```bash
> kubectl get configmap -n apisix apisix -oyaml
apiVersion: v1
kind: ConfigMap
  name: apisix
  namespace: apisix
data:
  config.yaml: |-
......
      stream_proxy:             
        only: false
        tcp:                       
          - addr: 9100
            tls: true
          - addr: 9200
          - addr: 9300
            tls: true
......
```

### Service
```bash
kubectl get service -n apisix apisix-gateway -o yaml
apiVersion: v1
kind: Service
metadata:
  name: apisix-gateway
  namespace: apisix
......
  - name: proxy-tcp-0
    nodePort: 31062
    port: 9100
    protocol: TCP
    targetPort: 9100
  - name: proxy-tcp-1
    nodePort: 32600
    port: 9200
    protocol: TCP
    targetPort: 9200
  - name: proxy-tcp-2
    nodePort: 32367
    port: 9300
    protocol: TCP
    targetPort: 9300
......
```

**Review by tokers — APPROVED** [review]

**Review by tao12345666333 — APPROVED** [review]
LGTM

**Pull request merged**